src/http/request/curl.rs

```diff
@@ -0,0 +1,127 @@
+use curl::easy::{Easy, List, Transfer};
+use http_types::header::HeaderValue;
+use http_types::Method;
+
+fn headermap_to_curl_list(headermap: &http_types::HeaderMap<HeaderValue>) -> List {
+    let mut list = List::new();
+    headermap.iter().for_each(|(k, v)| {
+        // this will scan for printable US-ASCII only bytes
+        let header = v
+            .to_str()
+            .map(|hval| [k.as_str(), ": ", hval].concat())
+            .expect("found header value without a displayable US-ASCII string");
+        list.append(header.as_str())
+            .expect("failed to allocate node for curl list of headers");
+    });
+    list
+}
+
+#[derive(Debug)]
+pub enum CurlEasyClient<'easy, 'data> {
+    Easy(&'easy Easy),
+    Transfer(Transfer<'easy, 'data>),
+}
+
+impl<'easy, 'data> From<Transfer<'easy, 'data>> for CurlEasyClient<'easy, 'data> {
+    fn from(t: Transfer<'easy, 'data>) -> Self {
+        Self::Transfer(t)
+    }
+}
+
+impl<'easy> From<&'easy Easy> for CurlEasyClient<'easy, '_> {
+    fn from(e: &'easy Easy) -> Self {
+        Self::Easy(e)
+    }
+}
+
+impl<'easy, 'data> CurlEasyClient<'easy, 'data> {
+    pub fn new(easy: &'easy Easy) -> Self {
+        Self::Easy(easy)
+    }
+
+    pub fn perform(&self) -> Result<(), curl::Error> {
+        match self {
+            Self::Easy(e) => e.perform(),
+            Self::Transfer(t) => t.perform(),
+        }
+    }
+
+    pub fn into_transfer(self) -> Option<Transfer<'easy, 'data>> {
+        match self {
+            Self::Transfer(t) => Some(t),
+            Self::Easy(_) => None,
+        }
+    }
+
+    pub fn easy(&self) -> Option<&'easy Easy> {
+        if let Self::Easy(e) = self {
+            Some(e)
+        } else {
+            None
+        }
+    }
+}
+#[cfg(feature = "curl-easy")]
+impl<'easy, 'data, URI: ToString> FromRequest<(&'easy mut Easy, URI)>
+    for CurlEasyClient<'easy, 'data>
+{
+    fn from_request(r: Request, params: (&'easy mut Easy, URI)) -> Self {
+        let (uri, body) = r.parameters.uri_and_body(r.path);
+        let (client, uri_base) = params;
+        let uri = uri_base.to_string() + uri.as_ref();
+
+        match r.method {
+            Method::GET => client.get(true),
+            Method::POST => client.post(true),
+            Method::PUT => client.put(true),
+            // any other verb needs to use custom_request()
+            m => client.custom_request(m.as_str()),
+        }
+        .expect("failed to set up the request's method");
+
+        client
+            .url(uri.as_str())
+            .expect("error setting up url for curl");
+        let mut headerlist = headermap_to_curl_list(&r.headers);
+        // libcurl by default adds "Expect: 100-continue" to send bodies, which would break us
+        headerlist
+            .append("Expect:")
+            .expect("failed to allocate node for curl list of headers");
+        // don't specify Content-Type for this request (similar to other clients)
```

Comment:
?

Reply:
Hard to not see that. :)

This is optional for apisonator. The semantics would be "this uses the default body encoding", although the docs mention that it in reality is the usual content type for post forms (which would be what libcurl would use if this wasn't being used). See https://github.com/3scale/apisonator/blob/a99715d10f5816908bdd4184f8c48663728a4255/docs/rfcs/error_responses.md#currently-known-error_codes-and-proposed-classification under `error -> content_type_invalid`.

The main issue here is something I wanted to actively avoid dealing with for a while until we had advanced a bit more elsewhere, but since you raised this... The problem is whether we want to actually go through the encoding of the full set or parameters or not (that is, _really_ encoding it as opposed to just having a header), as historical usage seems to show that we have never hit a problem with plain text using the query format (ie. no values have odd and ambiguous characters like `&`, though I suspect they could have, especially for things like `app_key`).

This is confirmed by code in a few other clients. The Ruby one does escaping of service id and service credentials (provider key and service token), but not app id, not app key (which it completely fails to even include in reports), nor the full usage data and payload in general when using a body in the report call: https://github.com/3scale/3scale_ws_api_for_ruby/blob/c4952601a3ad3b6c3e5fda6dffc003bbcaaca3aa/lib/3scale/client.rb#L178-L185.

A similar problem arises when the same data is being used by auth*.xml in the query parameters of `GET` requests. Any such field values _and_ keys (including metric names) should not have any values that would need to be URL-escaped to work correctly, or would otherwise need to be URL-escaped. That is, `app_key=1&2&3` cannot possibly work for authorizations unless URL-escaped, because all authorizations use `GET` requests and these parameters are sent in as URL query parameters. Then taking a look at this again, it turns out that the only client that does this correctly is the Ruby one (this time for everything in the URL rather than just service credentials).

If we want to go this route (which I think is the correct thing to do), we might want to make it more explicit in Apisonator docs and encode the parameters already like this at a higher level when we build them in our client, maybe thinking whether it would make sense to just use escaped values all the time - and, well, de facto declare most other clients as broken, but you could argue they all already are under specific circumstances.

Reply:
I see. I think that the only parameter that might be affected by this are application credentials. See this issue and the JIRA it references: https://github.com/3scale/pisoni/pull/22

I don't think it's worth it to tackle this at this stage of the project, but definitely something to keep in mind for the future.